src/storage/in_mem_csv_copier/in_mem_arrow_node_copier.cpp

```diff
@@ -0,0 +1,591 @@
+                                               node_offset_t startOffset, uint64_t numValues) {
+        for (auto i = 0u; i < numValues; i++) {
+            auto offset = i + startOffset;
+            if constexpr (is_same<T, int64_t>::value) {
+                auto key = (int64_t *) column->getElement(offset);
+                if (!pkIndex->append(*key, offset)) {
+                    throw CopyCSVException(Exception::getExistedPKExceptionMsg(to_string(*key)));
+                }
+            } else {
+                auto element = (ku_string_t *) column->getElement(offset);
+                auto key = column->getInMemOverflowFile()->readString(element);
+                if (!pkIndex->append(key.c_str(), offset)) {
+                    throw CopyCSVException(Exception::getExistedPKExceptionMsg(key));
+                }
+            }
+        }
+    }
+
+    template<typename T1, typename T2>
+    arrow::Status InMemArrowNodeCopier::batchPopulateColumnsTask(uint64_t primaryKeyPropertyIdx,
+                                                                 uint64_t blockId,
+                                                                 uint64_t offsetStart,
+                                                                 HashIndexBuilder<T1> *pkIndex,
+                                                                 InMemArrowNodeCopier *copier,
+                                                                 const vector<shared_ptr<T2>> &batchColumns,
+                                                                 char delimiter) {
+        copier->logger->trace("Start: path={0} blkIdx={1}", copier->csvDescription.filePath, blockId);
+        vector<PageByteCursor> overflowCursors(copier->nodeTableSchema->getNumStructuredProperties());
+        for (auto blockOffset = 0u; blockOffset < copier->numLinesPerBlock[blockId]; ++blockOffset) {
+            putPropsOfLineIntoColumns(copier->structuredColumns,
+                                      overflowCursors,
+                                      batchColumns,
+                                      offsetStart + blockOffset,
+                                      blockOffset,
+                                      delimiter);
+        }
+        populatePKIndex(copier->structuredColumns[primaryKeyPropertyIdx].get(), pkIndex, offsetStart,
+                        copier->numLinesPerBlock[blockId]);
+        copier->logger->trace("End: path={0} blkIdx={1}", copier->csvDescription.filePath, blockId);
+        return arrow::Status::OK();
+    }
+
+    template<typename T>
+    void InMemArrowNodeCopier::putPropsOfLineIntoColumns(
+            vector<unique_ptr<InMemColumn>> &structuredColumns, vector<PageByteCursor> &overflowCursors,
+            const std::vector<shared_ptr<T>> &arrow_columns, uint64_t nodeOffset, uint64_t blockOffset,
+            char delimiter) {
+        for (auto columnIdx = 0u; columnIdx < structuredColumns.size(); columnIdx++) {
```

Comment:
As we discussed, we should try to find a way that can directly copy from an arrow column to the corresponding kuzu column. Copy values from arrow columns to kuzu columns row by row seems inefficient. Besides, we should try to avoid converting from arrow value to string, then back to cpp primitive types when possible.
Can you open an issue after you merge the PR? We can solve these altogether after you guys have made changes to the rel csv copier.